janus-idp/operator
# Update CRD with meaningful set of parameters

## Description
This updates the CRD structure based on the proposal and discussion in https://github.com/janus-idp/operator/issues/21#issuecomment-1834055506, later refined in https://docs.google.com/document/d/14z9tE2XsCRl0hLTbHkyOdv26S6trHLgfpxYOfo-cm1I/edit

This is an example of what we agreed to have:

```yaml
spec:
  skipLocalDb: false
  application:
    appConfig:
      # All ConfigMaps will be mounted in the same place
      mountPath: "/opt/app-root/src"
      configMaps:
          - name: "my-app-config-cm12"
            # key is optional
            key: "my-file1"
    extraFiles:
      # All ConfigMaps and Secrets will be mounted in the same place
      mountPath: /opt/app-root/src/path/to/my-extra-files
      configMaps:
          - name: "my-extra-config-secret12"
            # key is optional. All files will be mounted if not set
            key: "my-file1"
      secrets:
          - name: "my-extra-config-secret12"
            # key is optional. All files will be mounted if not set
            key: "my-file1"    
    # CM must have a key named 'dynamic-plugins.yaml'
    dynamicPluginsConfigMapName: "my-dynamic-plugins-config-cm"

    extraEnvs:
      configMaps:
        - name: "cm-name"
          # key is optional. Env vars from all keys if key is not set
          key: "key"
      secrets:
        - name: "secret-name"
          # key is optional. Env vars from all keys if key is not set
          key: "key"
      envs:
        - name: MY_ENV_VAR_1
          value: my-value-1
    replicas: 2
    image: "quay.io/rhdh/rhdh-hub-rhel9:1.0-194"
    imagePullSecrets:
      - rhdh-pull-secret
```

More context in:
- https://github.com/janus-idp/operator/issues/21
- https://github.com/janus-idp/operator/issues/26
- https://docs.google.com/document/d/14z9tE2XsCRl0hLTbHkyOdv26S6trHLgfpxYOfo-cm1I/edit

## Which issue(s) does this PR fix or relate to

- Fixes #26 
- Fixes https://github.com/janus-idp/operator/issues/52

## PR acceptance criteria


## How to test changes / Special notes to the reviewer

The most [basic CR (with no spec)](https://github.com/janus-idp/operator/blob/main/examples/bs1.yaml) should work.

See examples of more complete CRs here: 
- [`janus-cr.yaml`](https://github.com/janus-idp/operator/pull/56/files#diff-2b2b184d4fa7f0bbe12cb1f7343c29cd11efe020113e6589d34f8db087c98590)
- [`rhdh-cr.yaml`](https://github.com/janus-idp/operator/pull/56/files#diff-97ad27a95157aa628d70d5b55bf7de03ac3066a42af25d698dacc8f8a24ffc11)

**Noteworthy changes**:
- app-config files can be mounted under the same mount path (`/opt/app-root/src` by default)
 - extra-config files can be mounted under the same mount path (`/opt/app-root/src` by default)
- Because dynamic plugins configuration needs to be handled differently from app-configs, the field introduced in https://github.com/janus-idp/operator/pull/27 is still present. It is now renamed into a simple `dynamicPluginsConfigMapName` string field.
- The `backendAuthSecretRef` field introduced in https://github.com/janus-idp/operator/pull/27 has been removed to make it less confusing. Instead, if user did not specify any `appConfig`, `extraFiles` or `extraEnv` in their CR, we would create a default app-config with a default value for the backend auth key. Otherwise, users are expected to provide their own value for this key (either via their own app-config, or via an `APP_CONFIG_...` env var)
